nomos-mix/network/src/behaviour.rs

```diff
@@ -0,0 +1,231 @@
+    },
+    Multiaddr, PeerId,
+};
+use nomos_mix_message::{message_id, unwrap_message};
+
+use crate::{
+    error::Error,
+    handler::{FromBehaviour, MixConnectionHandler, ToBehaviour},
+};
+
+/// A [`NetworkBehaviour`] that forwards messages between mix nodes.
+pub struct Behaviour {
+    /// Peers that support the mix protocol, and their connection IDs
+    negotiated_peers: HashMap<PeerId, HashSet<ConnectionId>>,
+    /// Queue of events to yield to the swarm.
+    events: VecDeque<ToSwarm<Event, FromBehaviour>>,
+    /// An LRU time cache for storing seen messages (based on their ID). This cache prevents
+    /// duplicates from being propagated on the network.
+    duplicate_cache: TimedCache<Vec<u8>, ()>,
+}
+
+#[derive(Debug)]
+pub enum Event {
+    /// A fully unwrapped message received from one of the peers.
+    FullyUnwrappedMessage(Vec<u8>),
+    Error(Error),
+}
+
+impl Behaviour {
+    pub fn new() -> Self {
+        Self {
+            negotiated_peers: HashMap::new(),
+            events: VecDeque::new(),
+            duplicate_cache: TimedCache::with_lifespan(60),
+        }
+    }
+
+    /// Publishs a message through the mix network.
+    ///
+    /// This function expects that the message was already encoded for the cryptographic mixing
+    /// (e.g. Sphinx encoding).
+    ///
+    /// The message is forward to all connected peers,
+    /// so that it can arrive in the mix node who can unwrap it one layer.
+    /// Fully unwrapped messages are returned as the [`MixBehaviourEvent::FullyUnwrappedMessage`].
+    pub fn publish(&mut self, message: Vec<u8>) -> Result<(), Error> {
+        self.duplicate_cache.cache_set(message_id(&message), ());
+        self.forward_message(message, None)
+    }
+
+    /// Forwards a message to all connected peers except the one that was received from.
+    ///
+    /// Returns [`Error::NoPeers`] if there are no connected peers that support the mix protocol.
+    fn forward_message(
+        &mut self,
+        message: Vec<u8>,
+        propagation_source: Option<PeerId>,
+    ) -> Result<(), Error> {
+        let peer_ids = self
+            .negotiated_peers
+            .keys()
+            .filter(|&peer_id| {
+                if let Some(propagation_source) = propagation_source {
+                    *peer_id != propagation_source
+                } else {
+                    true
+                }
+            })
+            .cloned()
+            .collect::<Vec<_>>();
+
+        if peer_ids.is_empty() {
+            return Err(Error::NoPeers);
+        }
+
+        for peer_id in peer_ids.into_iter() {
+            tracing::debug!("Registering event for peer {:?} to send msg", peer_id);
+            self.events.push_back(ToSwarm::NotifyHandler {
+                peer_id,
+                handler: NotifyHandler::Any,
+                event: FromBehaviour::Message(message.clone()),
+            });
+        }
+
+        Ok(())
+    }
```

Comment:
If we do this manually. Why do we need `gossipsub` for anyway?

Reply:
We don't need `gossipsub` in this crate. The `gossipsub` is necessary only for the "broadcasting phase", which broadcasts fully-unwrapped messages (block proposals) to the network. Unlike the mix phase, the block proposals should (must) be delivered to all nodes in the network, as further as possible. For that, I think we need some practices that the gossipsub has, such as IHAVE/IWANT.
But, for the mix phase that this crate implements, we don't need `gossipsub`. A mix node forwards messages to all nodes that are currently connected to it. If the message was not delivered to a certain node because the node was offline, this crate do nothing to eventually deliver the message. Instead, the initial message sender (origin) should re-send the message.

Reply:
Ok. So my questions is. Does the broadcasting phase messages need to be scheduled as part of the mix protocol or not? I think I am missing something here.

Reply:
The broadcasting phase is totally independent from the mix phase, but is the part of the Nomos Mix protocol. In other words, the fully-unwrapped messages must be broadcasted (from their last mix node)

Reply:
Actually, that was why I implemented the composite network behaviour that has both `MixBehaviour` and `gossipsub::Behaviour`. But, I removed it because it can be done outside of this crate, even using a different swarm, as we discussed in the call.

Reply:
> The broadcasting phase is totally independent from the mix phase, but is the part of the Nomos Mix protocol. I’ll double check with Marcin whether the same network should be used for both phases.

I think we need to see this through. If we are forwarding messages manually here I do not think we need to have any gossipsub at all (because you are really doing the job here). If the gossiping is just the last phase in where the message is completely unwrapped we do not need gossiping from the mixnet behaviour or swarm. Let's have a chat about this in the proper channels. It feels important.

Reply:
True. That’s why the crate in this PR does NOT have the gossipsub behaviour. This crate implements only the MIX phase, not the broadcasting phase. I believe you noticed that this PR doesn’t use gossipsub at all. I’m just checking if we don’t have any misunderstanding. 

Whether to use gossipsub can be determined outside of this crate, if the broadcasting phase should be considered as a separate protocol (potentially using a different set of nodes). Of course, we have to clarify it before merging this PR. I started a thread about this topic in Discord.

Reply:
According to our discussion made in Discord, we can go with this crate which implements only the mix phase. In the upper layer, we can implement the broadcast phase using gossipsub, which broadcasts messages returned from this crate.